### ISSUE TITLE
Question/Request: Using float values with (sellupperpcnt, selllowerpcnt,trailingstoploss) params

### ISSUE THREAD

**Issue by github-ptaylor** [whittlem/pycryptobot#96]
hey @whittlem 

Can we use float values (i.e. .5) for (sellupperpct, selllowerpct,trailingstoploss)?

Most gains or losses happen in decimal percentages, so having the ability to use floats is more logical versus having to wait for a full  (1,2,3..) % gain or (-1)% loss to trigger a SELL.

the code logic when checking in the (config) area, looks for an INT:

`if 'sellupperpcnt' in config:
                            if isinstance(config['sellupperpcnt'], int):
                                if config['sellupperpcnt'] > 0 and config['sellupperpcnt'] <= 100:
                                    self.sell_upper_pcnt = int(config['sellupperpcnt'])`

the code logic when looking for command-line arguments looks for an INT and then tries to set it to FLOAT (that doesn't look right):

`if args.sellupperpcnt != None:
            if isinstance(args.sellupperpcnt, int):
                if args.sellupperpcnt > 0 and args.sellupperpcnt <= 100:
                    self.sell_upper_pcnt = float(args.sellupperpcnt)`

**Comment by github-ptaylor** [comment]
@whittlem @letsautom8  i made changes to the code to allow floats for those parameters, all in (PyCryptoboty.py).

Here is output from a successful  BUY/SELL with the changes:

--------------------------------------------------------------------------------
|                Market : ALICEUSDT                                            |
|           Granularity : 5m                                                   |
--------------------------------------------------------------------------------
|              Bot Mode : LIVE - live trades using your funds!                 |
|           Bot Started : 2021-05-01 10:06:16.958378                           |
================================================================================
|        Sell Upper : 0.5%                                                     |
|        Sell Lower : -0.5%                                                    

—

================================================================================
|         Iteration : 1                                                        |
|         Timestamp : 2021-05-01 16:05:00                                      |
--------------------------------------------------------------------------------
|             Close : 17.71                                                    |
|             EMA12 : 17.28                                                    |
|             EMA26 : 17.25                                                    |
|    Crossing Above : True                                                     |
|   Currently Above : True                                                     |
|    Crossing Below : False                                                    |
|   Currently Below : False                                                    |
|         Condition : EMA12 is currently crossing above EMA26                  |
|             SMA20 : 17.17                                                    |
|            SMA200 : 16.85                                                    |
--------------------------------------------------------------------------------
|              MACD : 0.03                                                     |
|            Signal : -0.05                                                    |
|   Currently Above : True                                                     |
|   Currently Below : False                                                    |
|         Condition : MACD is currently above Signal and has crossed over      |
--------------------------------------------------------------------------------
|            Action : BUY                                                      |
================================================================================
--------------------------------------------------------------------------------
|                      *** Executing LIVE Buy Order ***                        |
--------------------------------------------------------------------------------
ALICE balance before order: 0.0
USDT balance before order: 84.5069

—

================================================================================
|         Iteration : 1                                                        |
|         Timestamp : 2021-05-01 16:05:00                                      |
--------------------------------------------------------------------------------
|             Close : 0.84                                                     |
|             EMA12 : 0.84                                                     |
|             EMA26 : 0.84                                                     |
|    Crossing Above : False                                                    |
|   Currently Above : False                                                    |
|    Crossing Below : False                                                    |
|   Currently Below : True                                                     |
|         Condition : EMA12 is currently below EMA26 and has crossed over      |
|             SMA20 : 0.84                                                     |
|            SMA200 : 0.78                                                     |
--------------------------------------------------------------------------------
|              MACD : -0.01                                                    |
|            Signal : -0.01                                                    |
|   Currently Above : False                                                    |
|   Currently Below : True                                                     |
|         Condition : MACD is currently below Signal and has crossed over      |
--------------------------------------------------------------------------------
|            Action : WAIT                                                     |
================================================================================
Order quantity after rounding and fees: 0.95
ALICE balance after order: 0.95
USDT balance after order: 67.6151

—

================================================================================
|         Iteration : 2                                                        |
|         Timestamp : 2021-05-01 16:10:00                                      |
--------------------------------------------------------------------------------
|             Close : 0.15                                                     |
|             EMA12 : 0.15                                                     |
|             EMA26 : 0.15                                                     |
|    Crossing Above : False                                                    |
|   Currently Above : False                                                    |
|    Crossing Below : False                                                    |
|   Currently Below : True                                                     |
|         Condition : EMA12 is currently below EMA26 and has crossed over      |
|             SMA20 : 0.15                                                     |
|            SMA200 : 0.16                                                     |
--------------------------------------------------------------------------------
|              MACD : -0.01                                                    |
|            Signal : -0.01                                                    |
|   Currently Above : True                                                     |
|   Currently Below : False                                                    |
|         Condition : MACD is currently above Signal and has crossed over      |
--------------------------------------------------------------------------------
|            Action : WAIT                                                     |
================================================================================
! Trailing Stop Loss Triggered (< -0.5%) 

—

================================================================================
|         Iteration : 2                                                        |
|         Timestamp : 2021-05-01 16:10:00                                      |
--------------------------------------------------------------------------------
|             Close : 17.56                                                    |
|             EMA12 : 17.31                                                    |
|             EMA26 : 17.26                                                    |
|    Crossing Above : False                                                    |
|   Currently Above : True                                                     |
|    Crossing Below : False                                                    |
|   Currently Below : False                                                    |
|         Condition : EMA12 is currently above EMA26 and has crossed over      |
|             SMA20 : 17.17                                                    |
|            SMA200 : 16.84                                                    |
--------------------------------------------------------------------------------
|              MACD : 0.04                                                     |
|            Signal : -0.04                                                    |
|   Currently Above : True                                                     |
|   Currently Below : False                                                    |
|         Condition : MACD is currently above Signal and has crossed over      |
--------------------------------------------------------------------------------
|            Action : SELL                                                     |
================================================================================
|            Margin : -0.87%%                                                  |
================================================================================
--------------------------------------------------------------------------------
|                      *** Executing LIVE Sell Order ***                        |
--------------------------------------------------------------------------------
ALICE balance before order: 0.95
USDT balance before order: 67.6151
Order quantity after rounding and fees: 0.94
ALICE balance after order: 0.01
USDT balance after order: 84.0749

I'll show you the changes i made in a follow-up post

**Comment by github-ptaylor** [comment]
Here is the snippet of the output of a diff on my code compared to the main branch:

`                         if 'sellupperpcnt' in config:
-                            if isinstance(config['sellupperpcnt'], int):
-                                if config['sellupperpcnt'] > 0 and config['sellupperpcnt'] <= 100:
-                                    self.sell_upper_pcnt = int(config['sellupperpcnt'])
+                            if isinstance(config['sellupperpcnt'], (int,str)):
+                                if float(config['sellupperpcnt']) > 0 and float(config['sellupperpcnt'] <= 100):
+                                    self.sell_upper_pcnt = float(config['sellupperpcnt'])
 
                         if 'selllowerpcnt' in config:
-                            if isinstance(config['selllowerpcnt'], int):
-                                if config['selllowerpcnt'] >= -100 and config['selllowerpcnt'] < 0:
-                                    self.sell_lower_pcnt = int(config['selllowerpcnt'])
+                            if isinstance(config['selllowerpcnt'], (int,str)):
+                                if float(config['selllowerpcnt']) >= -100 and float(config['selllowerpcnt']) < 0:
+                                    self.sell_lower_pcnt = float(config['selllowerpcnt'])
 
                         if 'trailingstoploss' in config:
-                            if isinstance(config['trailingstoploss'], int):
-                                if config['trailingstoploss'] >= -100 and config['trailingstoploss'] < 0:
-                                    self.tailing_stop_loss = int(config['trailingstoploss'])
+                            if isinstance(config['trailingstoploss'], (int,str)):
+                                if float(config['trailingstoploss']) >= -100 and float(config['trailingstoploss']) < 0:
+                                    self.tailing_stop_loss = float(config['trailingstoploss'])`

--

`         if args.sellupperpcnt != None:
-            if isinstance(args.sellupperpcnt, int):
-                if args.sellupperpcnt > 0 and args.sellupperpcnt <= 100:
+            if isinstance(args.sellupperpcnt, (int,str)):
+                if float(args.sellupperpcnt) > 0 and float(args.sellupperpcnt) <= 100:
                     self.sell_upper_pcnt = float(args.sellupperpcnt)
 
         if args.selllowerpcnt != None:
-            if isinstance(args.selllowerpcnt, int):
-                if args.selllowerpcnt >= -100 and args.selllowerpcnt < 0:
+            if isinstance(args.selllowerpcnt, (int,str)):
+                if float(args.selllowerpcnt) >= -100 and float(args.selllowerpcnt) < 0:
                     self.sell_lower_pcnt = float(args.selllowerpcnt)
 
         if args.trailingstoploss != None:
-            if isinstance(args.trailingstoploss, int):
-                if args.trailingstoploss >= -100 and args.trailingstoploss < 0:
+            if isinstance(args.trailingstoploss, (int,str)):
+                if float(args.trailingstoploss) >= -100 and float(args.trailingstoploss) < 0:
                     self.trailing_stop_loss = float(args.trailingstoploss)`

**Comment by github-ptaylor** [comment]
Here is snippet of my code changes:

         if args.sellupperpcnt != None:
-            if isinstance(args.sellupperpcnt, int):
-                if args.sellupperpcnt > 0 and args.sellupperpcnt <= 100:
+            if isinstance(args.sellupperpcnt, (int,str)):
+                if float(args.sellupperpcnt) > 0 and float(args.sellupperpcnt) <= 100:
                     self.sell_upper_pcnt = float(args.sellupperpcnt)

**Comment by letsautom8** [comment]
Please create a PR :)

**Comment by github-ptaylor** [comment]
> Would u mind creating a PR pls? Any improvements are welcome.

hey @letsautom8 , sure, i am reading up on how to do all that.

**Comment by whittlem** [comment]
I reviewed the PR which looks good but the validation needs some adjusting. If you can fix that I'll approve and include it. Thanks for the contribution.
